src/test/kotlin/com/ampnet/identityservice/persistence/repository/AutoInvestTaskRepositoryTest.kt

```diff
@@ -0,0 +1,281 @@
+package com.ampnet.identityservice.persistence.repository
+
+import com.ampnet.identityservice.TestBase
+import com.ampnet.identityservice.persistence.model.AutoInvestTask
+import com.ampnet.identityservice.persistence.model.AutoInvestTaskHistoryStatus
+import com.ampnet.identityservice.persistence.model.AutoInvestTaskStatus
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
+import org.springframework.test.context.junit.jupiter.SpringExtension
+import java.math.BigInteger
+import java.time.ZoneId
+import java.time.ZonedDateTime
+import java.util.UUID
+
+@DataJpaTest
+@ExtendWith(value = [SpringExtension::class])
+@AutoConfigureTestDatabase
+class AutoInvestTaskRepositoryTest : TestBase() {
+
+    @Autowired
+    private lateinit var autoInvestTaskRepository: AutoInvestTaskRepository
+
+    @Test
+    fun mustCorrectlyInsertAutoInvestTask() {
+        val task = taskWithAmount(BigInteger.valueOf(123456789L))
+
+        suppose("auto-invest task is inserted into database") {
+            val numModified = autoInvestTaskRepository.createOrUpdate(task)
+            assertThat(numModified).isOne()
+        }
+
+        verify("auto-invest task was correctly inserted into database") {
+            val databaseTask = autoInvestTaskRepository.findById(task.uuid)
+            assertThat(databaseTask).hasValue(task)
+        }
+    }
+
+    @Test
+    fun mustCorrectlyUpdateExistingAutoInvestTask() {
+        val task = taskWithAmount(BigInteger.valueOf(500L))
+
+        suppose("auto-invest task is inserted into database") {
+            val numModified = autoInvestTaskRepository.createOrUpdate(task)
+            assertThat(numModified).isOne()
+        }
+
+        val newTask = taskWithAmount(BigInteger.valueOf(1_000L))
+
+        suppose("another auto-invest task for the same user and campaign is inserted into database") {
+            val numModified = autoInvestTaskRepository.createOrUpdate(newTask)
+            assertThat(numModified).isOne()
+        }
+
+        verify("auto-invest task is correctly updated") {
+            val databaseTask = autoInvestTaskRepository.findById(task.uuid)
+            assertThat(databaseTask).hasValue(
+                task.copy(
+                    amount = task.amount + newTask.amount,
+                    createdAt = newTask.createdAt
+                )
+            )
```

Comment:
@filipdujmusic @mightymatth Updating auto-invest amount should result with the task that has updated new amount(new/last sent value) not the sum of those two? Setting allowance always overrides the old value?

Reply:
I think that it should be the sum old old and new values when auto-invest didn't happen yet, at least that is what I remember we talked about.

Reply:
The amount in task always has to reflect the the amount approved by the user. That's the ultimate source of truth in any given moment.

#### SCENARIO ####
> User has started task and approved $500.
> 500$ task is still not being processed an the user goes to invest another $1000. The user manages to quickly execute this investment before the previous task was processed (highly unlikely since the user has to go through the full ramp flow and wait for the usdc funds to arrive on his wallet,and all of this has to occur before our script has processed the last investment)
> on the blockchain side this means that the approval was *overwritten* and the approval is now $1000 (not $500 as before - regardless of the fact that the user has bought total of 1500 USDC)
> this means that our script can only take $1000 of user funds and process it as investment

Based on this, I thnk we should override tasks if they start coming in quickly for the same (investor, campaign) pair. Or do the check for everytime the task is received , read the approval amount from the chain and make sure there exists only one (investor, campaign) item in the queue and it's amount is  configured to the value read from the chain. Does this make sense to you guys?

Reply:
Ok, approve will be called from the frontend after the user has signed the allowance transaction, each allowance sets the new value, overrides the old one. The same has to be in our task, if the task for combination(investor, campaign) is not completed and the new amount is received, we should override the amount in the task. Sum of these values will always be reverted in the smart contact because it amount will be smaller of set allowance.

Reply:
OK, makes sense to me, I will make it override the value instead of using the sum of new and old ones.